packages/engine/Source/Scene/Model/MetadataPipelineStage.js

```diff
@@ -450,9 +457,30 @@ function addPropertyTexturePropertyMetadata(renderResources, propertyInfo) {
     metadataVariable
   );
 
-  // Get a GLSL expression for the value of the property
+  // Get a GLSL expression for the texture coordinates of the property.
+  // By default, this will be taken directly from the attributes.
   const texCoordVariable = `attributes.texCoord_${texCoord}`;
-  const valueExpression = `texture(${textureUniformName}, ${texCoordVariable}).${channels}`;
+  let texCoordVariableExpression = texCoordVariable;
+
+  // Check if the texture defines a `transform` from a `KHR_texture_transform`
+  if (defined(transform) && !Matrix3.equals(transform, Matrix3.IDENTITY)) {
+    // Add a uniform for the transformation matrix
+    const transformUniformName = `${textureUniformName}Transform`;
+    shaderBuilder.addUniform(
+      "mat3",
+      transformUniformName,
+      ShaderDestination.FRAGMENT
+    );
+    uniformMap[transformUniformName] = function () {
+      return transform;
+    };
+
+    // Update the expression for the texture coordinates
+    // with one that transforms the texture coordinates
+    // with the transform matrix first
+    texCoordVariableExpression = `vec2(${transformUniformName} * vec3(${texCoordVariable}, 1.0))`;
```

Comment:
@javagl to answer your question, I think this inlining is a fine way to implement this extension for property textures (and feature ID textures will be similar).

for the `MaterialPipelineStage`, the use of `#define`s was because the possible textures (base color, normal, etc.) are all known in advance according to the glTF spec.

Meanwhile, metadata is much more variable in the sense that the quantity and names of uniforms depends on the number of properties. This is why this pipeline stage uses a code generation approach.